idaes/models_extra/power_generation/unit_models/soc_submodels.py

```diff
@@ -0,0 +1,3997 @@
+
+        @self.Expression(tset, iznodes)
+        def current(b, t, iz):
+            return b.current_density[t, iz] * b.xface_area[iz]
+
+        @self.Expression(tset, ixnodes, iznodes)
+        def voltage_drop(b, t, ix, iz):
+            return b.current[t, iz] * b.resistance[t, ix, iz]
+
+        @self.Expression(tset, iznodes)
+        def resistance_total(b, t, iz):
+            return sum(b.resistance[t, ix, iz] for ix in ixnodes)
+
+        @self.Expression(tset, iznodes)
+        def voltage_drop_total(b, t, iz):
+            return sum(b.voltage_drop[t, ix, iz] for ix in ixnodes)
+
+        @self.Constraint(tset, ixnodes, iznodes, comps)
+        def material_balance_eqn(b, t, ix, iz, i):
+            return b.node_volume[ix, iz] * b.dcdt[t, ix, iz, i] == b.xface_area[iz] * (
+                b.xflux[t, ix, iz, i] - b.xflux[t, ix + 1, iz, i]
+            ) + b.zface_area[ix] * (b.zflux[t, ix, iz, i] - b.zflux[t, ix, iz + 1, i])
+
+        if dynamic:
+            self.material_balance_eqn[tset.first(),:,:,:].deactivate()
+            
+        @self.Expression(tset, ixnodes, iznodes)
+        def joule_heating(b, t, ix, iz):
+            return b.current[t, iz] * b.voltage_drop[t, ix, iz]
+
+        @self.Constraint(tset, ixnodes, iznodes)
+        def energy_balance_solid_eqn(b, t, ix, iz):
+            return (
+                b.node_volume[ix, iz] * (b.porosity * b.dcedt[t, ix, iz]
+                 + (1-b.porosity) * b.dcedt_solid[t, ix, iz])
+                == b.xface_area[iz] * (b.qxflux[t, ix, iz] - b.qxflux[t, ix + 1, iz])
+                + b.zface_area[ix] * (b.qzflux[t, ix, iz] - b.qzflux[t, ix, iz + 1])
+                + b.joule_heating[t, ix, iz]
+                # For mass flux heat transfer include exchange with channel
+                # probably make little differece, but want to ensure the energy
+                # balance closes
+                + b.xface_area[iz]
+                * sum(
+                    b.xflux[t, ix, iz, i]
+                    * comp_enthalpy_expr(b.temperature_xfaces[t, ix, iz], i)
+                    for i in comps
+                )
+                - b.xface_area[iz]
+                * sum(
+                    b.xflux[t, ix + 1, iz, i]
+                    * comp_enthalpy_expr(b.temperature_xfaces[t, ix + 1, iz], i)
+                    for i in comps
+                )
+                + b.zface_area[ix]
+                * sum(
+                    b.zflux[t, ix, iz, i]
+                    * comp_enthalpy_expr(b.temperature_zfaces[t, ix, iz], i)
+                    for i in comps
+                )
+                - b.zface_area[ix]
+                * sum(
+                    b.zflux[t, ix, iz + 1, i]
+                    * comp_enthalpy_expr(b.temperature_zfaces[t, ix, iz + 1], i)
+                    for i in comps
+                )
+            )
+        if dynamic:
+            self.energy_balance_solid_eqn[tset.first(),:,:].deactivate()
+    def initialize_build(
```

Comment:
Missing line break. I would suggest running `black` on all this code before the PR is merged - this is going to be enforced in the near future.